### PR TITLE
feat: add optional id and clv to JWT claims

### LINKED ISSUE #6016: Feature request: IDs for JWT claims
### Problem description

The execution API spec defines an optional `id` claim that can be set in a JWT token to convey information about the client being authenticated. This is useful for more advanced forms of authentication beyond the basic 1-CL-to-1-EL authentication that we usually do. For example, my [_Electric Eel_](https://github.com/sigp/eleel) multiplexer uses the `id` to identify the connecting client and decide which JWT secret to use. Without the ID, it has to fall back on iterating all known JWT secrets and trying to verify the token against them (which is slow).

The spec defines the `id` claim here: https://github.com/ethereum/execution-apis/blob/main/src/engine/authentication.md#jwt-claims

It's worth noting that the ID in the claim is _distinct_ from the key-ID concept defined by the JWT standard, which is part of the _header_, not the body. Apart from being a bit inelegant this isn't an issue.

### Solution description

Add a flag to Lodestar like `--jwt-id` which takes a string to use as the JWT ID for claims made to the EL.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by ensi321**
**Motivation**

When running multiple CL clients with different JWT secrets, it is difficult for EL clients to identify which JWT secrets to verify against which could be solved by having identifiers such as `id` and `clv` included in the tokens sent to the EL clients. 

**Description**

- Add `id: string` and `clv: string` to JwtClaim, which are used in `JsonRpcHttpClient`
- Add `jwtId` and `jwtVersion` to eth1 and execution engine options
- Introduce `--jwtId` cli flag to pass the value to `jwtCliam.id`. 
- Auto-populate `jwtVersion` in eth1 and execution engine opts during beacon handler init which are used in `jwtClaim.clv`
- Rename `--jwt-secret` to `--jwtSecret` for the sake of being consistent with naming convention of other flags. Note this change is backward compatible since yargs handles hyphen case and camel case automatically

They are compliant with the [execution-api](https://github.com/ethereum/execution-apis/blob/main/src/engine/authentication.md#jwt-claims) specs.

Closes #6016

### CODE REVIEW

File: packages/beacon-node/src/eth1/provider/eth1Provider.ts [hunk 1]
```diff
@@ -76,6 +78,8 @@ export class Eth1Provider implements IEth1Provider {
       // Don't fallback with is truncated error. Throw early and let the retry on this class handle it
       shouldNotFallback: isJsonRpcTruncatedError,
       jwtSecret: opts.jwtSecretHex ? fromHex(opts.jwtSecretHex) : undefined,
+      jwtId: opts.jwtId ?? undefined,
+      jwtVersion: opts.jwtVersion ?? undefined,
```
Comment (nflaig) [inline packages/beacon-node/src/eth1/provider/eth1Provider.ts]: Why is this required? Omitting `?? undefined` should have the same result

```suggestion
      jwtId: opts.jwtId,
      jwtVersion: opts.jwtVersion,
```

File: packages/cli/src/options/beaconNodeOptions/execution.ts [hunk 2]
```diff
@@ -80,4 +84,18 @@ export const options: CliCommandOptions<ExecutionEngineArgs> = {
...
+  "jwt-id": {
+    description:
+      "An optional identifier to be included in the id field in the claims of the jwt tokens for authentication with EL client's rpc server hosting engine apis",
+    type: "string",
+    group: "execution",
+  },
+
+  "jwt-version": {
+    description:
+      "An option version string to be included in the clv field in the claims of the jwt tokens for authentication with EL client's rpc server hosting engine apis",
```
Comment (nflaig) [inline packages/cli/src/options/beaconNodeOptions/execution.ts]: I don't get the use case of this

- Why would a user set a arbitrary value here, and not set the actual version like how it is done for libp2p agent version and user-agent header?
https://github.com/ChainSafe/lodestar/blob/c50db8f7ae47198c76070487a6cfd0009d9bfd39/packages/cli/src/cmds/beacon/handler.ts#L199-L201
- Even if we assume this is a boolean value to set it or not, what's the point. As I see it, we either send this all the time or not at all. This is used to tell the EL about what CL is calling it, maybe to track it via metrics or add some kind of temporary workaround to fix a client bug or misbehavior (this is the use case for builder user agent as far as I know). Now, if it is off by default, both those use case don't work.

Maybe I am missing the purpose of CLV but could also just consider to add it all the time and only disable it if `--private` flag is set
https://github.com/ChainSafe/lodestar/blob/c50db8f7ae47198c76070487a6cfd0009d9bfd39/packages/cli/src/cmds/beacon/options.ts#L119-L120

File: packages/cli/src/options/beaconNodeOptions/execution.ts [hunk 3]
```diff
@@ -80,4 +84,18 @@ export const options: CliCommandOptions<ExecutionEngineArgs> = {
     type: "string",
     group: "execution",
   },
+
+  "jwt-id": {
```
Comment (nflaig) [inline packages/cli/src/options/beaconNodeOptions/execution.ts]: I'd suggest we are not adding more flags using this format and start using camelCase for consistency with other CLI flags

```suggestion
  "jwtId": {
```

Maybe it makes sense to also rename the `jwt-secret` flag to `jwtSecret` while keeping previous value as an alias, all docs should reference new flag name

**Review by nflaig — COMMENTED** [review]
Just some initial thoughts

File: packages/cli/src/options/beaconNodeOptions/execution.ts (hunk 2 above)
Comment (ensi321) [reply packages/cli/src/options/beaconNodeOptions/execution.ts]: > I don't get the use case of this
> 
> * Why would a user set a arbitrary value here, and not set the actual version like how it is done for libp2p agent version and user-agent header?
>   https://github.com/ChainSafe/lodestar/blob/c50db8f7ae47198c76070487a6cfd0009d9bfd39/packages/cli/src/cmds/beacon/handler.ts#L199-L201
> * Even if we assume this is a boolean value to set it or not, what's the point. As I see it, we either send this all the time or not at all. This is used to tell the EL about what CL is calling it, maybe to track it via metrics or add some kind of temporary workaround to fix a client bug or misbehavior (this is the use case for builder user agent as far as I know). Now, if it is off by default, both those use case don't work.

https://github.com/ChainSafe/lodestar/blob/c50db8f7ae47198c76070487a6cfd0009d9bfd39/packages/cli/src/cmds/beacon/options.ts#L119-L120

I am having the same doubt as you. I think having a user configurable version provides flexibility. Also the [spec](https://github.com/ethereum/execution-apis/blob/main/src/engine/authentication.md#jwt-claims) describes `clv` as **type/version** and not just version. 

I looked at lighthouse for reference, and they are just having a flag for [jwt-version](https://github.com/sigp/lighthouse/blob/stable/beacon_node/src/config.rs#L353). No mechanism to auto-populate this value. This behaviour of `id` and `clv` in this PR mimics the current behaviour of LH

> 
> Maybe I am missing the purpose of CLV but could also just consider to add it all the time and only disable it if `--private` flag is set
> 
> 
I also think the current use case of `clv` is quite limited, and it could be a bit wasteful to add it all the time. I prefer to have it enabled only when user needs it

Comment (nflaig) [reply packages/cli/src/options/beaconNodeOptions/execution.ts]: > I had the same doubt as you as well. I think having a user configurable version provides flexibility.

What the difference to the ID then?

I guess right now CLV is unused at least as far I am aware but based on what the spec says, setting an arbitrary value sounds wrong
> The consensus layer client MAY use this to communicate the consensus layer client type/version.

Comment (ensi321) [reply packages/cli/src/options/beaconNodeOptions/execution.ts]: > > I had the same doubt as you as well. I think having a user configurable version provides flexibility.
> 
> What the difference to the ID then?
> 
> I guess right now CLV is unused at least as far I am aware but based on what the spec says, setting an arbitrary value sounds wrong
> 
> > The consensus layer client MAY use this to communicate the consensus layer client type/version.

I've reached out to LH team on discord about this, and imo their answer isn't too convincing.

I think we can have our own interpretation on the spec instead of mimicking LH's. 

How about we have `--include-jwt-id` and `--include-jwt-version` flags? If user passes in a value after the flag, Lodestar will include `id`/`clv` with the mentioned value. If user only adds the flag without any value, we populate default value eg. "" for `id`, and `Lodestar/${version}` for `clv`?

Comment (nflaig) [reply packages/cli/src/options/beaconNodeOptions/execution.ts]: Thanks for checking with them. I would just go with an implementation that supports current use cases and can be easily extended without introduction breaking changes in the future (i.e. not have to rename CLI flags).

Let's look at `id` first
- we know it is used by [eleel](https://github.com/sigp/eleel)
- the value is used to map id to a secret, see [eleel#getting-started](https://github.com/sigp/eleel#getting-started)
- to do the mapping, it needs to be a static and known value that is set by a user on both clients

I can't think of a use case where you would want a random id, not known to the server. In fact, the jwt standard has another property for this use case, see ["jti" (JWT ID) Claim](https://datatracker.ietf.org/doc/html/rfc7519#section-4.1.7).

Regarding the name of the flag, `--include-jwt-id` sounds a bit like it would expect a boolean value, so I'd rather go with `--jwt-id`.

For the `clv` the only reasonable approach to me is if Lodestar itself sets it to `Lodestar/${version}` as we do for libp2p and builder requests. Could be behind a feature flag but I still don't see what problem we are trying to solve at this point, once there is an actual use case for this could add it with a clear understanding on how it has to work.

> If user only adds the flag without any value, we populate default value

I don't think we have this behavior implemented on any other CLI flags (flag acting as boolean & string) and I am not sure if it worth to add complexity for a use case we don't even know exists.

Comment (wemeetagain) [reply packages/cli/src/options/beaconNodeOptions/execution.ts]: I think we can just include the clv all the time. I haven't benchmarked it, but likely adding these few extra bytes in the payload will have a negligible effect.

As far as the id is concerned, using `--jwt-id` or better `--jwtId` would match the naming of the existing jwt option. We probably shouldn't include an id in the payload unless the user has configured one.

File: packages/cli/src/options/beaconNodeOptions/execution.ts (hunk 3 above)
Comment (wemeetagain) [reply packages/cli/src/options/beaconNodeOptions/execution.ts]: yargs automatically populates args with both hyphen case and camel case, kinda an automagic feature iirc. So this can be done and it won't be breaking.

File: packages/cli/src/options/beaconNodeOptions/execution.ts (hunk 2 above)
Comment (nflaig) [reply packages/cli/src/options/beaconNodeOptions/execution.ts]: > As far as the id is concerned, using --jwt-id would match the naming of the existing jwt option.

We have changed flag names in the past and kept previous value as an alias, we could even have `jwt-id` as an alias for `jwtId` if we want to.

The only concern I can see if we rename `jwt-secret` to `jwtSecret` is if someone wants to look up what it does in the docs. I can't find a strong argument against renaming while having consistent naming patterns is not only cleaner UX but also makes them more intuitive, i.e. you can guess what the flag name might be without looking at docs, right now this is not the case for `jwt-secret`

> I think we can just include the clv all the time. I haven't benchmarked it, but likely adding these few extra bytes in the payload will have a negligible effect.

+1, I'd say either that or not include at all

File: packages/cli/src/options/beaconNodeOptions/execution.ts (hunk 3 above)
Comment (nflaig) [reply packages/cli/src/options/beaconNodeOptions/execution.ts]: That's good to know, if it doesn't do that we can always just add an alias
https://github.com/ChainSafe/lodestar/blob/4a25589900c660bd37b15b3be8bedf0672d69340/packages/cli/src/cmds/validator/options.ts#L173

Comment (ensi321) [reply packages/cli/src/options/beaconNodeOptions/execution.ts]: > I'd suggest we are not adding more flags using this format and start using camelCase for consistency with other CLI flags
> 
> Maybe it makes sense to also rename the `jwt-secret` flag to `jwtSecret` while keeping previous value as an alias, all docs should reference new flag name

Indeed yargs automatically handles hyphen and camel case. I will use `jwtId` here.

I will also rename `jwt-secret` to `jwtSecret` in this PR since it's a small change and wouldn't break anything

File: packages/cli/src/options/beaconNodeOptions/execution.ts (hunk 2 above)
Comment (ensi321) [reply packages/cli/src/options/beaconNodeOptions/execution.ts]: > > As far as the id is concerned, using --jwt-id would match the naming of the existing jwt option.
> 
> We have changed flag names in the past and kept previous value as an alias, we could even have `jwt-id` as an alias for `jwtId` if we want to.
> 
> The only concern I can see if we rename `jwt-secret` to `jwtSecret` is if someone wants to look up what it does in the docs. I can't find a strong argument against renaming while having consistent naming patterns is not only cleaner UX but also makes them more intuitive, i.e. you can guess what the flag name might be without looking at docs, right now this is not the case for `jwt-secret`
> 
> > I think we can just include the clv all the time. I haven't benchmarked it, but likely adding these few extra bytes in the payload will have a negligible effect.
> 
> +1, I'd say either that or not include at all

Updated!

File: docs/usage/beacon-management.md [hunk 4]
```diff
@@ -33,7 +33,12 @@ You must generate a secret 32-byte (64 characters) hexadecimal string that will
 
 ### Configure Lodestar to locate the JWT secret
 
-When starting up a Lodestar beacon node in any configuration, ensure you add the `--jwt-secret $JWT_SECRET_PATH` flag to point to the saved secret key file.
+When starting up a Lodestar beacon node in any configuration, ensure you add the `--jwtSecret $JWT_SECRET_PATH` flag to point to the saved secret key file.
+
+### Set up and include identifiers in JWT tokens
+
+Lodestar auto-populates `clv` field in the claims of JWT authentication tokens with a non-configurable value `Lodestar/$CLIENT_VERSION` eg. `Lodestar/v1.3.0/2d0938e` to communicate the client's version. Lodestar also optionally includes `id` field in the claims with value `$JWT_ID` if the appropriate flag `--jwtId $JWT_ID` is added. 
+`id` and `clv` are particularly useful when running multiple consensus-layer clients with the different JWT secrets which makes the execution-layer client difficult to choose which JWT secret to verify against due to the inability to distinguish between the different consensus-layer clients.
```
Comment (nflaig) [inline docs/usage/beacon-management.md]: This might be confusing, as far as I am aware execution clients do not natively support multiplexing of consensus clients and I don't think this will ever be something that will be implemented. Without a multiplexer middleware like [eleel](https://github.com/sigp/eleel) it is not recommended to connect multiple CLs to a single EL.

File: packages/beacon-node/src/eth1/provider/jsonRpcHttpClient.ts [hunk 5]
```diff
@@ -103,6 +105,11 @@ export class JsonRpcHttpClient implements IJsonRpcHttpClient {
        * and it might deny responses to the RPC requests.
        */
       jwtSecret?: Uint8Array;
+      /** If jwtSecret and jwtId are provided, jwtId will be included in JwtClaim.id */
+      jwtId?: string;
+      /** If jwtSecret is provided, jwtVersion will be included in JwtClaim.clv. 
+       *  jwtVersion has default value of empty string */
+      jwtVersion?: string;
```
Comment (nflaig) [inline packages/beacon-node/src/eth1/provider/jsonRpcHttpClient.ts]: - I think it would be better to omit JwtClaim.clv if jwtVersion is not set
- jsdoc formatting

File: packages/beacon-node/src/eth1/provider/jsonRpcHttpClient.ts [hunk 6]
```diff
@@ -255,7 +264,13 @@ export class JsonRpcHttpClient implements IJsonRpcHttpClient {
          *
          * Jwt auth spec: https://github.com/ethereum/execution-apis/pull/167
          */
-        const token = encodeJwtToken({iat: Math.floor(new Date().getTime() / 1000)}, this.jwtSecret);
+        const jwtClaim: JwtClaim = {
+          iat: Math.floor(new Date().getTime() / 1000),
```
Comment (nflaig) [inline packages/beacon-node/src/eth1/provider/jsonRpcHttpClient.ts]: Let's simplify this

```suggestion
          iat: Math.floor(Date.now() / 1000),
```

File: packages/beacon-node/src/eth1/provider/jwt.ts [hunk 7]
```diff
@@ -4,8 +4,10 @@ import jwt from "jwt-simple";
 
 const {encode, decode} = jwt;
 
-/** jwt token has iat which is issued at unix timestamp, and an optional exp for expiry */
-type JwtClaim = {iat: number; exp?: number};
+/** jwt token has iat which is issued at unix timestamp, an optional exp for expiry,
+ *  an optional id for unique identifier, and an optional clv for client type/version
```
Comment (nflaig) [inline packages/beacon-node/src/eth1/provider/jwt.ts]:
```suggestion
 *  an optional id as unique identifier, and an optional clv for client type/version
```

File: packages/beacon-node/src/execution/engine/http.ts [hunk 8]
```diff
@@ -55,6 +55,14 @@ export type ExecutionEngineHttpOpts = {
    * +-5 seconds interval.
    */
   jwtSecretHex?: string;
+  /**
+   * A identifier string passed by user in cli arg that will be included in `id` field in jwt claims
```
Comment (nflaig) [inline packages/beacon-node/src/execution/engine/http.ts]:
```suggestion
   * An identifier string passed as CLI arg that will be set in `id` field of jwt claims
```

CLI all uppercase seems most consistent with existing code, ideally, in comments should probably also write JWT instead of jwt (does not seem to be case in existing comments, but could consider updating those in this PR)

File: packages/beacon-node/src/execution/engine/http.ts [hunk 9]
```diff
@@ -55,6 +55,14 @@ export type ExecutionEngineHttpOpts = {
...
+   */
+  jwtId?: string;
+  /**
+   * A version string passed by user in cli arg that will be included in `clv` field in jwt claims
```
Comment (nflaig) [inline packages/beacon-node/src/execution/engine/http.ts]:
```suggestion
   * A version string that will be set in `clv` field of jwt claims
```

File: packages/cli/src/cmds/beacon/handler.ts [hunk 10]
```diff
@@ -199,6 +199,10 @@ export async function beaconHandlerInit(args: BeaconArgs & GlobalArgs) {
     beaconNodeOptions.set({network: {version: version.split("/")[0]}});
     // Add User-Agent header to all builder requests
     beaconNodeOptions.set({executionBuilder: {userAgent: `Lodestar/${version}`}});
+    // Set jwt version with version string
+    beaconNodeOptions.set({executionEngine: {jwtVersion: `Lodestar/${version}`}});
+    // Set jwt version with version string
+    beaconNodeOptions.set({eth1: {jwtVersion: `Lodestar/${version}`}});
```
Comment (nflaig) [inline packages/cli/src/cmds/beacon/handler.ts]: Could also consider using a const for `Lodestar/${version}`, instead of repeating it 3 times

```suggestion
    // Set jwt version with version string
    beaconNodeOptions.set({executionEngine: {jwtVersion: `Lodestar/${version}`}, eth1: {jwtVersion: `Lodestar/${version}`}});
```

File: packages/cli/src/options/beaconNodeOptions/execution.ts [hunk 11]
```diff
@@ -74,10 +76,17 @@ export const options: CliCommandOptions<ExecutionEngineArgs> = {
...
+  "jwtSecret": {
     description:
       "File path to a shared hex-encoded jwt secret which will be used to generate and bundle HS256 encoded jwt tokens for authentication with the EL client's rpc server hosting engine apis. Secret to be exactly same as the one used by the corresponding EL client.",
     type: "string",
     group: "execution",
   },
+
+  "jwtId": {
+    description:
+      "An optional identifier to be included in the id field in the claims of the jwt tokens for authentication with EL client's rpc server hosting engine apis",
```
Comment (nflaig) [inline packages/cli/src/options/beaconNodeOptions/execution.ts]:
```suggestion
      "An optional identifier to be set in the id field of the claims included in jwt tokens for authentication with EL client's rpc server hosting engine apis",
```

File: docs/usage/beacon-management.md (hunk 4 above)
Comment (nflaig) [inline docs/usage/beacon-management.md]: Just a side note, I'd rather see use the terms "consensus client" and "execution client" in the docs. This seems to be the more conventional terminology and it's easier to read / less verbose

File: packages/beacon-node/test/unit/eth1/jwt.test.ts [hunk 12]
```diff
@@ -10,6 +10,14 @@ describe("ExecutionEngine / jwt", () => {
     expect(decoded).toEqual(claim);
   });
 
+  it("encode/decode correctly with id and clv", () => {
+    const jwtSecret = Buffer.from(Array.from({length: 32}, () => Math.round(Math.random() * 255)));
+    const claim = {iat: Math.floor(new Date().getTime() / 1000), id: "4ac0", clv: "v1.11.3"};
```
Comment (nflaig) [inline packages/beacon-node/test/unit/eth1/jwt.test.ts]: As tests can serve as documentation of code behavior a lot of time, could consider using a proper clv value (as Lodestar would include it)

File: docs/usage/beacon-management.md (hunk 4 above)
Comment (ensi321) [reply docs/usage/beacon-management.md]: > Just a side note, I'd rather see use the terms "consensus client" and "execution client" in the docs. This seems to be the more conventional terminology and it's easier to read / less verbose

We can have a separate PR to systemically update the term "consensus-layer client" and "execution-layer client"👍

Comment (ensi321) [reply docs/usage/beacon-management.md]: > This might be confusing, as far as I am aware execution clients do not natively support multiplexing of consensus clients and I don't think this will ever be something that will be implemented. Without a multiplexer middleware like [eleel](https://github.com/sigp/eleel) it is not recommended to connect multiple CLs to a single EL.

I thought about the possible use case of configuring `--jwtId` for normal users and I can't think of any. The multiplexer case is nice but users who read this doc probably won't ever need it. 

I am inclined to just remove this entirely because the point of this doc is to set up a beacon node. Explaining things in jwt claim and the option to add `id` because of multiplexer seem really out of place in the doc.

Man writing user doc is hard

File: docs/usage/beacon-management.md [hunk 13]
```diff
@@ -33,7 +33,12 @@ You must generate a secret 32-byte (64 characters) hexadecimal string that will
 
 ### Configure Lodestar to locate the JWT secret
 
-When starting up a Lodestar beacon node in any configuration, ensure you add the `--jwt-secret $JWT_SECRET_PATH` flag to point to the saved secret key file.
+When starting up a Lodestar beacon node in any configuration, ensure you add the `--jwtSecret $JWT_SECRET_PATH` flag to point to the saved secret key file.
```
Comment (dapplion) [inline docs/usage/beacon-management.md:36]: Why change this flag? I think yargs supports both notations and the `--jwt-secret` flag is kind of standarized across clients

Comment (nflaig) [reply docs/usage/beacon-management.md:36]: > kind of standarized across clients

Kinda, but not really
- Teku: `--ee-jwt-secret-file` ([ref](https://github.com/dappnode/DAppNodePackage-teku/blob/master/beacon-chain/entrypoint.sh#L37))
- Lighthouse: `--execution-jwt` ([ref](https://github.com/dappnode/DAppNodePackage-lighthouse/blob/master/beacon-chain/entrypoint.sh#L51))
- Prsym: `--jwt-secret` ([ref](https://github.com/dappnode/DAppNodePackage-prysm/blob/main/beacon-chain/entrypoint.sh#L64))
- Nimbus: `--jwt-secret` ([ref](https://github.com/dappnode/DAppNodePackage-nimbus/blob/main/build/entrypoint.sh#L70))

For Lodestar, `--jwt-secret` is currently the only commonly used flag which uses different casing. I'd recommend we use the same standard (camelCase) for this flag as well.

File: docs/usage/beacon-management.md (hunk 4 above)
Comment (nflaig) [reply docs/usage/beacon-management.md]: > The multiplexer case is nice but users who read this doc probably won't ever need it.

Yeah, I rather see people using eleel come to the CLI reference to see if Lodestar has the option to set the jwt id

> I am inclined to just remove this entirely because the point of this doc is to set up a beacon node. 

+1, I think should keep it simple here

> Man writing user doc is hard

Defintiely hard thing because it is quite subjective and user feedback is required in a lot of cases, @matthewkeil is working on a docs restructuring which should make it clearer on what details to include where.

In my opinion
- step-by-step instructions should mostly omit unnecessary details
- while sections explaining different concepts can go more into detail

File: packages/cli/src/cmds/beacon/handler.ts [hunk 14]
```diff
@@ -195,10 +195,14 @@ export async function beaconHandlerInit(args: BeaconArgs & GlobalArgs) {
...
+    const versionStr = `Lodestar/${version}`;
+    const simpleVersionStr = version.split("/")[0];
     // Add simple version string for libp2p agent version
-    beaconNodeOptions.set({network: {version: version.split("/")[0]}});
+    beaconNodeOptions.set({network: {version: simpleVersionStr}});
     // Add User-Agent header to all builder requests
-    beaconNodeOptions.set({executionBuilder: {userAgent: `Lodestar/${version}`}});
+    beaconNodeOptions.set({executionBuilder: {userAgent: versionStr}});
+    // Set jwt version with version string
+    beaconNodeOptions.set({executionEngine: {jwtVersion: versionStr}, eth1: {jwtVersion: versionStr}});
```
Comment (nflaig) [inline packages/cli/src/cmds/beacon/handler.ts:205]: Maybe should update the description of `--private` flag and mention this as well

https://github.com/ChainSafe/lodestar/blob/18dc0c958189cb60fd8df7db6b1ebdf060edbd96/packages/cli/src/cmds/beacon/options.ts#L119-L120

I guess at some point need to have more general description if there are more things we omit if node is set to private

File: packages/beacon-node/src/eth1/provider/jwt.ts [hunk 15]
```diff
@@ -4,8 +4,10 @@ import jwt from "jwt-simple";
 
 const {encode, decode} = jwt;
 
-/** jwt token has iat which is issued at unix timestamp, and an optional exp for expiry */
-type JwtClaim = {iat: number; exp?: number};
+/** jwt token has iat which is issued at unix timestamp, an optional exp for expiry,
+ *  an optional id as unique identifier, and an optional clv for client type/version
+ */
+export type JwtClaim = {iat: number; exp?: number; id?: string; clv?: string};
```
Comment (nflaig) [inline packages/beacon-node/src/eth1/provider/jwt.ts]:
```suggestion
/** 
 *  jwt token has iat which is issued at unix timestamp, an optional exp for expiry,
 *  an optional id as unique identifier, and an optional clv for client type/version
 */
export type JwtClaim = {iat: number; exp?: number; id?: string; clv?: string};
```

File: packages/cli/src/options/beaconNodeOptions/execution.ts [hunk 16]
```diff
@@ -74,10 +76,17 @@ export const options: CliCommandOptions<ExecutionEngineArgs> = {
...
+  "jwtSecret": {
     description:
       "File path to a shared hex-encoded jwt secret which will be used to generate and bundle HS256 encoded jwt tokens for authentication with the EL client's rpc server hosting engine apis. Secret to be exactly same as the one used by the corresponding EL client.",
     type: "string",
     group: "execution",
   },
+
+  "jwtId": {
+    description:
+      "An optional identifier to be set in the id field of the claims included in jwt tokens for authentication with EL client's rpc server hosting engine apis",
```
Comment (nflaig) [inline packages/cli/src/options/beaconNodeOptions/execution.ts]:
```suggestion
      "An optional identifier to be set in the id field of the claims included in jwt tokens used for authentication with EL client's rpc server hosting engine apis",
```

File: packages/beacon-node/src/eth1/provider/jwt.ts [hunk 17]
```diff
@@ -4,8 +4,11 @@ import jwt from "jwt-simple";
 
 const {encode, decode} = jwt;
 
-/** jwt token has iat which is issued at unix timestamp, and an optional exp for expiry */
-type JwtClaim = {iat: number; exp?: number};
+/**
+ *  jwt token has iat which is issued at unix timestamp, an optional exp for expiry,
+ *  an optional id as unique identifier, and an optional clv for client type/version
```
Comment (nflaig) [inline packages/beacon-node/src/eth1/provider/jwt.ts]:
```suggestion
 * jwt token has iat which is issued at unix timestamp, an optional exp for expiry,
 * an optional id as unique identifier, and an optional clv for client type/version
```

**Review by nflaig — DISMISSED** [review]
lgtm, thanks for all the updates

**Comment by wemeetagain** [comment]
@naviechan can you confirm that `--jwt-secret` still works and we don't need to set an alias?

**Comment by ensi321** [comment]
> @naviechan can you confirm that `--jwt-secret` still works and we don't need to set an alias?

Yes both `--jwt-secret` and `--jwtSecret` are correctly parsed

**Review by wemeetagain — DISMISSED** [review]

**Review by nflaig — APPROVED** [review]

**Review by wemeetagain — APPROVED** [review]

**Comment by wemeetagain** [comment]
:tada: This PR is included in [v1.12.0](https://github.com/ChainSafe/lodestar/releases/tag/v1.12.0) :tada:
